metis-indexing/src/main/java/eu/europeana/indexing/fullbean/RdfToFullBeanConverter.java

```diff
@@ -63,15 +66,44 @@ public FullBeanImpl convertRdfToFullBean(RdfWrapper record) {
         europeanaAggregation.map(new EuropeanaAggregationFieldInput()).orElse(null));
     europeanaAggregation.map(EuropeanaAggregationType::getCompleteness).map(LiteralType::getString)
         .map(Integer::parseInt).ifPresent(fullBean::setEuropeanaCompleteness);
-    fullBean.setTimestampCreated(europeanaAggregation.map(EuropeanaAggregationType::getCreated)
-        .map(Created::getString).map(RdfToFullBeanConverter::convertToDate).orElse(null));
-    fullBean.setTimestampUpdated(europeanaAggregation.map(EuropeanaAggregationType::getModified)
-        .map(Modified::getString).map(RdfToFullBeanConverter::convertToDate).orElse(null));
+    fullBean.setTimestampCreated(
+        europeanaAggregation.map(EuropeanaAggregationType::getCreated).map(Created::getString)
+            .map(RdfToFullBeanConverter::convertToDate).orElse(null));
+    fullBean.setTimestampUpdated(
+        europeanaAggregation.map(EuropeanaAggregationType::getModified).map(Modified::getString)
+            .map(RdfToFullBeanConverter::convertToDate).orElse(null));
 
     // Done.
     return fullBean;
   }
 
+  private List<Aggregation> convertAggregations(RdfWrapper record) {
+    final Supplier<List<WebResourceImpl>> webResourcesExtractor = new WebResourcesExtractor(record);
+    //Convert the provider aggregations
+    final List<AggregationImpl> providerAggregations = convertList(record.getProviderAggregations(),
+        new AggregationFieldInput(webResourcesExtractor), false);
+
+    //Compute the referenced web resources of the provider aggregation
+    final List<String> providerWebResourceReferences = Optional.ofNullable(providerAggregations)
+        .orElse(Collections.emptyList()).stream().map(AggregationImpl::getWebResources)
+        .flatMap(List::stream).map(WebResource::getAbout).collect(Collectors.toList());
+
+    //Create supplier for the web resources of the aggregator aggregation
+    final Supplier<List<WebResourceImpl>> aggregatorAggregationWebResources = () -> webResourcesExtractor
+        .get().stream().filter(
+            webResourceImpl -> !providerWebResourceReferences.contains(webResourceImpl.getAbout()))
+        .collect(Collectors.toList());
```

Comment:
I think there is a problem here, in your current implementation. I think that the code is such that all web resources are always added to the provider aggregation, so this supplier always returns an empty list, and empty web resources will be added to the aggregator aggregation if it references something.

I think the algorithm may be fixed by providing AggregationFieldInput with a lookup for webresources (by about value) instead of a supplier. And then make sure that it only adds what it references (taking care to remove web resources from the lookup if it has added it to an aggregation). Then you call this for all aggregations in the right order (provider aggregation first). And then at the end you  add all remaining (unclaimed) web resources to the (first) provider aggregation.

Does that sound reasonable?

Reply:
Yeah I see! I also see another issue with this..
The new implementation should cover also duplicate references for potential double provider aggregations.
Hope now it's better, plz have a look!